structuredheadings/plugin.js

```diff
@@ -1,35 +1,72 @@
-				}
-			});
-			
-			// Adding the button to the toolbar
-			editor.ui.addButton( 'structuredheadings', {
-				label: 'Autonumber Heading',
-				command: 'autoNumberHeading',
-				toolbar: 'styles,0'
-			})
-			return editor; // actually, no return is required, but this shuts eslint up
-		}
-	});
-	
-	function assignClassToList(list) {
-		// Iterate over the list of nodes
-		for (var i = 0; i < list.count(); i++) {
-			var h = list.getItem(i);
-			// Add the class if it doesn't have it
-			if(!h.hasClass('autonumber')) {
-				h.addClass('autonumber');
-			}
-		};
-	}
+  CKEDITOR.plugins.add("structuredheadings", {
+    icons: "structuredheadings",
+    init: function (editor) {
+      // list of elements allowed to be numbered
+      var allowedElements = ["h1", "h2", "h3", "h4", "h5", "h6"];
+
+      editor.addContentsCss(this.path + "styles/numbering.css");
+
+      // Helper function to set button state
+      var setState = function (state) {
+        switch (state) {
+        case "on":
+          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_ON);
+          break;
+        case "off":
+          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_OFF);
+          break;
+        case "disabled":
+          editor.getCommand("autoNumberHeading").setState(CKEDITOR.TRISTATE_DISABLED);
+          break;
+        }
+
+      };
+
+      // The command to add an autonumbering class to selection
+      editor.addCommand("autoNumberHeading", {
+        allowedContent: "h1(*); h2(*); h3(*); h4(*); h5(*); h6(*)",
+        requiredContent: allowedElements.join(";"),
+        startDisabled: true,
+        exec: function () {
+          var element = editor.getSelection().getStartElement();
+
+          if (!element.hasClass("autonumber")) {
+            element.addClass("autonumber");
+            setState("on");
+          } else {
+            element.removeClass("autonumber");
+            setState("off");
+          }
+        }
+      });
+
+      // Add the button to the toolbar only if toolbar plugin or button plugin is loaded
+      if (!!CKEDITOR.plugins.get("toolbar") || !!CKEDITOR.plugins.get("button")) {
+        editor.ui.addButton("structuredheadings", {
+          label: "Autonumber Heading",
+          command: "autoNumberHeading",
+          toolbar: "styles,0"
+        });
+      }
+
+      // Set button state on selection change.
+      // On/off for element style, disabled for invalid element
+      editor.on("selectionChange", function (e) {
+        var element = e.data.selection.getStartElement();
+        if (allowedElements.indexOf(element.$.localName.toString()) >= 0) {
+          if (element.hasClass("autonumber")) {
+            setState("on");
+          } else {
+            setState("off");
+          }
+        } else {
+          setState("disabled");
+        }
+
+      });
+
+      return editor; // actually, no return is required, but this shuts eslint up
```

Comment:
Yeah I really wasn't sure about this. You had it in your skeleton and I wasn't sure if it was going to still be required or not so I left it as is. But now I get why it was need, and so I've cleaned this up.